src/emqx_channel.erl

```diff
@@ -1698,14 +1704,25 @@ parse_topic_filters(TopicFilters) ->
     lists:map(fun emqx_topic:parse/1, TopicFilters).
 
 %%--------------------------------------------------------------------
-%% Ensure disconnected
+%% Maybe & Ensure disconnected
+
+ensure_disconnected(connected, Reason, Channel) ->
+    ensure_disconnected(Reason, Channel);
+ensure_disconnected(_, _, Channel) ->
+    Channel.
 
 ensure_disconnected(Reason, Channel = #channel{conninfo = ConnInfo,
                                                clientinfo = ClientInfo}) ->
     NConnInfo = ConnInfo#{disconnected_at => erlang:system_time(millisecond)},
-    ok = run_hooks('client.disconnected', [ClientInfo, Reason, NConnInfo]),
+    maybe_run_disconnected_hooks(Reason, [ClientInfo, Reason, NConnInfo]),
```

Comment:
Revert these changes and assert `maybe` in `ensure_disconnected/3`
We should keep the old behavior for other reasons like `keepalive_timeout `, `not_authorized`, `tcp_closed`, `internal_error`.

Reply:
`ensure_disconnected/3` is only for `Reason = takeovered` or `Reason = discarded`